### PR TITLE
 Introduce Semantics::visit_file_defs

### PR DESCRIPTION
**Pull request by Veykril**
See https://github.com/rust-analyzer/rust-analyzer/issues/3538#issuecomment-798920601

### CODE REVIEW

**Comment by matklad** [comment]
Can we write a (mark) test for this?

**Comment by matklad** [comment]
But yeah, a more appropriate fix here is to do less AST and more HIR. I think we might fail to crate hir for cfged-out items, and that perhaps can lead to similar issues?

I believe we already use hir for runnables, so we could even extract some geneic "visit all defs in the file" function.

**Comment by Veykril** [comment]
Ye I imagine this can still happen in some other cases, just that for files that aren't in the tree it will **always** happen. I can take a look at fixing it from another point of view later instead.

**Comment by Veykril** [comment]
Actually this is problematic, working with the HIR means that `file_structure` would have to work with the HIR, but with that the `Region` support that is being implemented in #7975 won't be possible anymore since the HIR doesnt track comments 😕 

And yep this also happens to cfg'ed items, so the fix here doesn't even work for that
grafik

**Comment by matklad** [comment]
Hm, but why file structure is at all relevant here? The fact that we use it
here is completely accidental and is wrong :)

That’s the standard mixing of two sides of abstraction: file structure is
intended for users outside of this crate. Using it as a private impl detail
of some other functionality adds unnecessary dependency between public API
and internal needs.

**Comment by Veykril** [comment]
@matklad ye I already noticed that myself a minute after I sent the comment, and edited my comment after that. Figured you might read this per mail which is therefor outdated 😅

File: crates/hir/src/semantics.rs [hunk 1]
```diff
@@ -155,6 +157,28 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
...
+                        .impl_defs(self.db)
+                        .into_iter()
+                        .for_each(|impl_| cb(Either::Right(impl_)));
+                }
+            }
+            cb(Either::Left(def));
+        }
+        module.impl_defs(self.db).into_iter().for_each(|impl_| cb(Either::Right(impl_)));
+    }
+
```
Comment (Veykril) [inline crates/hir/src/semantics.rs:181]: Is semantics the best spot to put this in? Was really unsure where this belongs

Comment (matklad) [reply crates/hir/src/semantics.rs:181]: I think this belongs to `ide_db` (or maybe just `ide`). I don't think semancis should be really aware about the division of code into files, it is IDE specific concern.

Comment (Veykril) [reply crates/hir/src/semantics.rs:181]: Ye that sounds like a better fit, moved it to `ide_db` now

**Comment by lnicola** [comment]
changelog fix make runnable detection more resilient to out-of-tree modules.

**Pull request merged**